src/quest/serializer.py

```diff
@@ -1,13 +1,41 @@
-from typing import Any, Dict, Tuple
-import importlib
+from typing import Any, Dict, Protocol, TypeVar, Callable
+from utils import get_obj_name, import_object
 
-from .utils import get_obj_name, import_object
+T = TypeVar('T')
 
 
-class MasterSerializer:
-    def __init__(self, type_serializers=None):
-        # Initialize with a dictionary of custom serializers - default to empty dict
+class StepSerializer(Protocol):
+    async def serialize(self, obj: Any) -> Any:
+        ...
+
+    async def deserialize(self, data: Any) -> Any:
+        ...
+
+
+class TypeSerializer(Protocol[T]):
+    async def serialize(self, obj: T) -> Dict[str, Any]:
+        ...
+
+    async def deserialize(self, data: Dict[str, Any]) -> T:
+        ...
+
+
+class NoopSerializer(StepSerializer):
+    async def serialize(self, obj: Any) -> Any:
+        # object already JSON-serializable
+        return obj
+
+    async def deserialize(self, data: Any) -> Any:
+        return data
+
+
+class MasterSerializer(StepSerializer):
+    def __init__(self, type_serializers: Dict[type, TypeSerializer[Any]] = None):
         self._type_serializers = type_serializers or {}
+        self._default_serializer = NoopSerializer()
+
+    def register_serializer(self, obj_type: type, serializer: TypeSerializer[Any]):
+        self._type_serializers[obj_type] = serializer
```

Comment:
I think it's safe to require all type serializers to be provided in the constructor (and thus delete this method)